src/Compilers/Core/MSBuildTask/ManagedCompiler.cs

```diff
@@ -278,6 +283,169 @@ protected override Encoding StandardOutputEncoding
+        /// Cancel the in-process build task.
+        /// </summary>
+        public override void Cancel()
+        {
+            base.Cancel();
+
+            _sharedCompileCts?.Cancel();
+        }
+
+        /// <summary>
+        /// Get the current directory that the compiler should run in.
+        /// </summary>
+        private string CurrentDirectoryToUse()
+        {
+            // ToolTask has a method for this. But it may return null. Use the process directory
+            // if ToolTask didn't override. MSBuild uses the process directory.
+            string workingDirectory = GetWorkingDirectory();
+            if (string.IsNullOrEmpty(workingDirectory))
+                workingDirectory = Environment.CurrentDirectory;
+            return workingDirectory;
+        }
+
+        /// <summary>
+        /// Get the "LIB" environment variable, or NULL if none.
+        /// </summary>
+        private string LibDirectoryToUse()
+        {
+            // First check the real environment.
+            string libDirectory = Environment.GetEnvironmentVariable("LIB");
+
+            // Now go through additional environment variables.
+            string[] additionalVariables = this.EnvironmentVariables;
+            if (additionalVariables != null)
+            {
+                foreach (string var in this.EnvironmentVariables)
+                {
+                    if (var.StartsWith("LIB=", StringComparison.OrdinalIgnoreCase))
+                    {
+                        libDirectory = var.Substring(4);
+                    }
+                }
+            }
+
+            return libDirectory;
+        }
+
+        /// <summary>
+        /// The return code of the compilation. Strangely, this isn't overridable from ToolTask, so we need 
+        /// to create our own.
+        /// </summary>
+        [Output]
+        public new int ExitCode { get; private set; } = 0;
+
+        /// <summary>
+        /// Handle a response from the server, reporting messages and returning
+        /// the appropriate exit code.
+        /// </summary>
+        private int HandleResponse(BuildResponse response)
+        {
+            var completedResponse = response as CompletedBuildResponse;
+            if (completedResponse != null)
+            {
+                LogMessages(completedResponse.Output, this.StandardOutputImportanceToUse);
+
+                if (LogStandardErrorAsError)
+                {
+                    LogErrorOutput(completedResponse.ErrorOutput);
+                }
+                else
+                {
+                    LogMessages(completedResponse.ErrorOutput, this.StandardErrorImportanceToUse);
+                }
+
+                ExitCode = completedResponse.ReturnCode;
+            }
+            else
+            {
+                Debug.Assert(response is MismatchedVersionBuildResponse);
+
+                LogErrorOutput("Roslyn compiler server reports different protocol version than build task.");
```

Comment:
Wouldn't it be useful to include the mismatched versions into the error message?